### ISSUE TITLE
twisted.internet.thread._putResultInDeferred should be public

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2845]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2845|
|Type|enhancement|
|Created|2007-10-07 17:31:05Z|
|Branch|https://github.com/twisted/twisted/tree/putresultindeferred-2845-4|

It's useful, and it's a common pattern when using twisted ```ThreadPool``` class. Furthermore, it's already used in adbapi.

Attachments:

* [twisted.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/29f/29f4249b905426195f314fdc042977a27ad7590c/d28a367b58eba790fb131d7a60d9534697fcdd09.patch) (6704 bytes) - added by davep on 2008-04-06 23:57:22Z - 
* [twisted-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/29f/29f4249b905426195f314fdc042977a27ad7590c/882cc950ef4f3c1a5a0094a8275bfdfb1f099f62.patch) (13685 bytes) - added by davep on 2008-04-19 17:55:12Z - 
* [twisted-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/29f/29f4249b905426195f314fdc042977a27ad7590c/56a15919362e98b3ec447bee64a64af396fb914d.patch) (20811 bytes) - added by davep on 2008-05-10 18:57:24Z - 
* [twisted-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/29f/29f4249b905426195f314fdc042977a27ad7590c/0c48d6603fb0ba6949c5a93fa939f6931ebed6b9.patch) (19018 bytes) - added by davep on 2008-06-08 00:38:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2845 2845
type__enhancement enhancement
reporter__therve therve
priority__normal normal
milestone__ 
branch__branches_putresultindeferred_2845_4 branches/putresultindeferred-2845-4
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1191778265000000 1191778265000000
changetime__1217850427000000 1217850427000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

I think it would be better to try to implement adbapi without making this public, rather than making it public because adbapi is implemented badly :).

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_2"></a>|@jdavisp3 commented|
|-|-|

So what would a better implementation of adbapi look like?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

An issue with the way threading is supported and the way `deferToThread` works (and hence `_putResultInDeferred`) is that it provides a `Deferred` with a result _before_ the worker thread which computed it has finished.  This means there will always be a race condition in code using this which can result in more threads than are necessary being created (that is, if all you ever do is call `deferToThread` once and then in a callback on that `Deferred` call `deferToThread` again, chances are you'll end up with two threads in your threadpool, rather than one).  This has minor negative performance consequences (more startup overhead, more memory used on some platforms), but more significant testing consequences, since there's no event that happens when everything is "clean".

So it'd be nice if `deferToThread were implemented differently.  Consequently `_putResultInDeferred` would probably disappear.  The replacement isn't totally obvious.  One solution involves changing IReactorThreads.  The way it is now, it's very hard to make a Deferred fire with the result of a threadpool task _after_ the thread which computed it is done.

So I haven't really answered the question, but maybe it's at least a bit more clear why the current thing isn't necessarily ideal and what problems a good solution would lack.

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_4"></a>|@jdavisp3 set owner to @jdavisp3<br>@jdavisp3 set status to `assigned`|
|-|-|

(Replying to [comment:3 exarkun])

Ok, thanks for clarifying that, and I have noticed that adbapi can make
more threads than it needs to. I will see if I can come up with a solution
to that.

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_5"></a>|@jdavisp3 commented|
|-|-|

So it seems that the issue is that _startSomeWorkers in twisted.python.threadpool
uses the length of self.working as part of its 'needed workers' calculation. And
a thread that is still executing its callback will be in that list.

So one solution is to push the split between 'run the function' and 'send back
the result' down into the threadpool so that the worker process can remove itself
from the list of working threads before sending the result using a non-blocking
callback.

I'm going to post a patch indicating a possible direction for a solution. It's
clearly not complete, but I would appreciate feedback on it's feasibility.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

The idea looks good, and it solves the problem reported by exarkun. But I think that the callback is a bit too dangerous:
 * the on_result callback is called in the thread. It's not immediatly obvious, and could lead to errors
 * the on_result callback should be as short as possible, because it falses the behavior of the threadpool: if it block, the thread is reported as available in the pool, whereas it's handling a task

So my suggestion is: why not simply pass a Deferred? I don't see an utility for something else, and that solves both problems.

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_7"></a>|@jdavisp3 commented|
|-|-|

Yeah, that's something I considered doing. I wasn't sure if it was
ok for stuff in twisted.python to 'know' about stuff in twisted.internet,
but it certainly doesn't bother me. I'll send another patch soon.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

Ooops.  You were right to be uncertain.  The dependency goes from `twisted.internet` to `twisted.python` - `twisted.python` can't use `twisted.internet` functionality.

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_9"></a>|@jdavisp3 commented|
|-|-|

Ah, well glad I didn't then :) I could split 'on_result' into 'on_result' and 'on_failure'
in the API so that clients could pass the callback and errback methods of a deferred, if
that would be better. Regardless, the need for the secondary callbacks to be non-blocking
will have to be documented clearly.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

Yeah I totally forgot this dependency problem. So maybe this is just a documentation issue: consenting adults, etc.

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_11"></a>|@jdavisp3 commented|
|-|-|

Ok, I will fill out this patch with docs & tests and we can see where we stand.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

[#3185](https://github.com/twisted/twisted/issues/3185) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_13"></a>|@jdavisp3 commented|
|-|-|

Ok, new patch with docs & tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

I think this is up for review?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [23410]) Branching to 'putresultindeferred-2845'

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_17"></a>|@jyknight commented|
|-|-|

Well....twisted/internet/defer clearly belongs in the twisted/python package. That it's not there is a historical mistake which is probably not worth fixing, but if it helps the API of this, it might be reasonable to pretend it is in twisted/python and call it okay to depend upon deferreds here...

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_18"></a>|@jdavisp3 commented|
|-|-|

It's up to you guys, I don't have an opinion. It's a straightforward change to make.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set owner to @jdavisp3|
|-|-|

  * twisted/test/test_threadpool.py
    * The `assert` methods are preferred over the `fail` methods now (and at some point we'll even write this down someplace ;)
    * Callbacks (except on Deferreds) typically shouldn't have assertions in them.  Instead they should collect some state and the assertion should be made later on.  This is to avoid the case where a bug causes the callback to not be called at all, making the test appear to succeed when it hasn't really tested what it was trying to test.
    * I feel like there should be something better than `_waitForLock` and `_waitForPool`, but I'm not sure what it is.
  * twisted/python/threadpool.py
    * `on_result` should be named `onResult`
    * `callInThread` is missing a docstring
    * `callInThreadWithCallback` is missing a docstring
    * The argument order for `callInThreadWithCallback` seems odd to me.  `on_result` is separating `func` from its arguments now.
    * There's no test coverage for the deletion of the `on_result` local in `callInThreadWithCallback` (see `test_threadCreationArguments`)
    * It might be good if everything that isn't needed is deleted before `on_result` is invoked and everything else is deleted afterwards.  Letting stuff get freed sooner is nice.  This should be tested, of course. ;)
    * There's no test coverage for which thread `on_result` is invoked in.
    * No test coverage for the context `on_result` is called with
  * twisted/internet/threads.py
    * `deferToThreadpool`'s nested function `on_result` doesn't need to take `d` as a parameter, it can just close over it.
    * No tests for `deferToThreadpool`?
  * twisted/internet/interfaces.py, twisted/internet/base.py
    * Hmm.  I'm not so sure about adding this method to `IReactorThreads`.  It almost seems as though we should deprecate all the methods on `IReactorThreads` and replace them with a `threadpool` attribute.  It'd be good to get some input from other people on this.

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_20"></a>|@jdavisp3 removed owner|
|-|-|

Ok, new patch posted addressing most of the review comments. I'm not sure
what's better than _waitForLock and _waitForPool, so I left them as is.

Instead of adding an identical method to IReactorThreads, I added 'getThreadpool'
and used that in threads.py. I didn't use an attribute because the reactor doesn't
actually create the threadpool until it needs to. Also, I didn't deprecate any of
the other IReactorThreads api calls.

This patch is relative to the trunk.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_21"></a>|@therve commented|
|-|-|

(In [23615]) Branching to 'putresultindeferred-2845-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve commented|
|-|-|

(In [23616]) Apply dave's patch

Refs [#2845](https://github.com/twisted/twisted/issues/2845)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_23"></a>|@therve commented|
|-|-|

(In [23617]) Make some trivial cleanups

Refs [#2845](https://github.com/twisted/twisted/issues/2845)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set owner to @jdavisp3|
|-|-|

  * twisted/enterprise/adbapi.py
    * `ConnectionPool.runWithConnection` could use a docstring
  * twisted/test/test_threads.py
    * test_deferredFailure docstring could be a bit more precise
    * `DeferToThreadpoolTestCase` is missing some docstrings and needs some name adjustments
  * twisted/test/test_threadpool.py
    * A number of the tests would be cleaner if they used `addCleanup` to get rid of their threadpool, rather than giant try/finally.
  * twisted/internet/threads.py
    * `deferToThreadpool` should take a reactor instead of importing the global one
  * twisted/internet/interfaces.py
    * docstring for `getThreadpool` should be a bit more descriptive, explain how it returns the threadpool used for the other `IReactorThreads` methods
  * there should be direct tests (in the style of those in `twisted.internet.test`) for the new reactor method (at least two, I'd say, since the implementation has a conditional in it)

Thanks.  I think we're getting close. :)

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_25"></a>|@jdavisp3 commented|
|-|-|

Ok, cool. This one will take me a little longer as I've managed to mess up my home network and need to wait for some new hardware.

**Comment by twisted-trac** [comment]
|[<img alt="jdavisp3's avatar" src="https://avatars.githubusercontent.com/u/43582?s=50" width="50" height="50">](https://github.com/jdavisp3)<a name="note_26"></a>|@jdavisp3 commented|
|-|-|

Ok, new patch. This one is relative to the tip of branches/putresultindeferred-2845-2.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

(In [24167]) apply twisted-4.patch

refs [#2845](https://github.com/twisted/twisted/issues/2845)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

(In [24168]) Branching to 'putresultindeferred-2845-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

Making some coding-standard changes

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_31"></a>|@exarkun commented|
|-|-|

(In [24179]) doc and coding standard changes

refs [#2845](https://github.com/twisted/twisted/issues/2845)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

(In [24180]) get rid of IReactorThreads.getThreadpool in favor of some abstraction violation in threads.py

refs [#2845](https://github.com/twisted/twisted/issues/2845)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_33"></a>|@exarkun commented|
|-|-|

(In [24378]) Branching to 'putresultindeferred-2845-4'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Please review.

Build results: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/putresultindeferred-2845-4

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_35"></a>|@radix set owner to @exarkun|
|-|-|

1.
```
+        @return: a Deferred which will fire the return value of
+            'func(Transaction(...), *args, **kw)', or a Failure.
```

C{} pls

2. I think it should be called "deferToThreadPool", since we call them "ThreadPool"s elsewhere, not "Threadpool"s.

3. Man, ThreadPool should really have a deferToThread method. Too bad deferred sucks.

4. ```+    def test_callInThreadWithCallbackException2(self):```

A better name for this test would be better.

5. ```+        Test C{callInThreadWithCallback} when onResult throws an exception.```

Please describe the behavior that is desired.

6. There are other cases like that.

7.
```
+    def test_singleThreadWithCallback(self):
+        """
+        Test that the creation of a new thread from the thread result callback
+        does not create an extra thread.
+        """
```

Can you explain this a little better? Is this a test for a bug that appeared during development? I don't think it's testing that "creation of a new thread" doesn't create an extra thread, but rather that calling callInThread does not create an extra thread.

8.

```
-        self.assertRaises(ValueError, pool.runWithConnection, raisingFunction)
+        d = pool.runWithConnection(raisingFunction)
```

Ugh, that old code was so dodgy. It didn't have the behavior of the real code at all! This is a great improvement.



Ok, I'm pretty happy with the refactoring being done here, the branch just needs to be cleaned up a bit.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun commented|
|-|-|

(In [24395]) better markup for runWithConnection docstring

refs [#2845](https://github.com/twisted/twisted/issues/2845)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_37"></a>|@exarkun commented|
|-|-|

(In [24396]) rename deferToThreadpool to deferToThreadPool

refs [#2845](https://github.com/twisted/twisted/issues/2845)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_38"></a>|@exarkun commented|
|-|-|

(In [24398]) Move the `test_singleThreadWithCallback test into `test_singleThread` and twiddle the docstring a bit

`test_singleThread` previously failed intermittently.  there is no way to make what it
was testing succeed reliably - that's why `onResult` is necessary for
`callInThreadWithCallback`; so change the test to use the API that can actually provide
this behavior reliably and drop the test for the unreliable because, because it's
unreliable.


refs [#2845](https://github.com/twisted/twisted/issues/2845)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_39"></a>|@exarkun commented|
|-|-|

(In [24399]) rewrite test_singleThread to omit the invalid multi-thread use of ThreadPool

refs [#2845](https://github.com/twisted/twisted/issues/2845)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_40"></a>|@exarkun removed owner|
|-|-|

Cleaned up, I think.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_41"></a>|@therve set owner to @exarkun|
|-|-|

 * Flake: ```twisted/enterprise/adbapi.py:11: 'defer' imported but unused```
 * deferToThreadPool documents its arguments with ```*args,**kwargs``` (same thing for runInteraction and runWithConnection) whereas deferToThread, callInThread, callInThreadWithCallback use ```args, kwargs```. I think the second group should be changed.

This looks great, please merge once both points fixed.

Build result here: http://buildbot.twistedmatrix.com/boxes-supported?branch=branches/putresultindeferred-2845-4&num_builds=1

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_42"></a>|@exarkun commented|
|-|-|

(In [24506]) remove unused import; change *arg, **kw doc style

refs [#2845](https://github.com/twisted/twisted/issues/2845)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_43"></a>|@exarkun set status to `closed`|
|-|-|

(In [24507]) Merge putresultindeferred-2845-4

Author: davep, exarkun
Reviewer: exarkun, therve
Fixes: [#2845](https://github.com/twisted/twisted/issues/2845)

Add `deferToThreadPool` to `twisted.internet.threads`.  This is similar
to `deferToThread`, but accepts a `ThreadPool` instance as a parameter
and uses it as the threadpool to which to dispatch calls.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
